File: November2021/C++/Largest_Component_Size_by_Common_Factor.cpp
class UnionFind{
    public : 
    UnionFind(int n)
    {
        // cout<<n<<endl;
        parent=vector<int>(n);
        for(int i=0;i<n;i++)
            parent[i]=i;
        
    }
    
    void unify(int i,int j)
    {
        int abosoluteparenti=getParent(i);
        int absoluteparentj=getParent(j);
        
        if(abosoluteparenti!=absoluteparentj)
        {
            parent[abosoluteparenti]=absoluteparentj;
        }
    }
    
    int getParent(int i)
    {
        // cout<<i<<endl;
        if(parent[i]==i)
            return i;
        parent[i]=getParent(parent[i]);
        return parent[i];
    }
    private :vector<int> parent;
    
};

class Solution {
public:
    int largestComponentSize(vector<int>& nums) {
        int val = *max_element(begin(nums),end(nums));
        UnionFind uf( val + 1);
        
        for(auto& i:nums)
        {
            for(int j=2;j<=sqrt(i);j++)
            {
                if(!(i%j))
                {
                    uf.unify(i, j);
                    uf.unify(i,i/j);
                }
            }
        }
        unordered_map<int,int> dict;
        int res=0;
        for(auto& i:nums)
        {
            int temp=uf.getParent(i);
            res=max(res,++dict[temp]);
            
            // res=dict[temp]>dict[res]?temp:res;
                 
        }
        // for(auto& i:dict) cout<<i.first<<" "<<i.second<<endl;
        return res;
    }
};

File: November2021/LargestComponentSizeByCommonFactor.java
class Solution {
	private class UnionFind {
		private int[] parent;

		private UnionFind(int n){
			parent = new int[n];
			for(int i=0;i<n;i++){
				parent[i] = i;
			}
		}
		private int getAbsoluteParent(int i){
			if(parent[i]==i){
				// absolute parent
				return i;
			}
			parent[i]=getAbsoluteParent(parent[i]);
			return parent[i];
		}

		private void unify(int i, int j){
			int absoluteParentI = getAbsoluteParent(i);
			int absoluteParentJ = getAbsoluteParent(j);
			if(absoluteParentI!=absoluteParentJ){
				// 7->2, 21->2 (7,21)
				parent[absoluteParentJ] = absoluteParentI;
			}
		}

	}

	public int largestComponentSize(int[] nums) {
		int max=0;
		for(int el:nums) {
			max = Math.max(el, max);
		}
		UnionFind uf = new UnionFind(max+1);

		for(int i:nums) {
			for(int j=2;j*j<=i;j++) {
				if(i%j==0){ // j is a factor of i
					uf.unify( j,i); // i=21, j=3
					uf.unify(i/j,i); // (21,3), (21,21/3(7))
				}
			}
		}
		//<Parent, Freq>
		Map<Integer, Integer> map = new HashMap<Integer, Integer>();
		int maxComponet = 0;
		for(int el:nums){
			int parent = uf.getAbsoluteParent(el);
			map.put(parent, map.getOrDefault(parent, 0) + 1);
			maxComponet= Math.max(maxComponet, map.get(parent));
		}



		return maxComponet;
	}
}